cpp/src/arrow/ipc/reader.cc

```diff
@@ -891,11 +905,12 @@ Status ReadSparseCSRIndex(const flatbuf::SparseTensor* sparse_tensor, int64_t nd
   RETURN_NOT_OK(
       file->ReadAt(indices_buffer->offset(), indices_buffer->length(), &indices_data));
 
-  std::vector<int64_t> indptr_shape({ndim + 1});
+  std::vector<int64_t> indptr_shape({shape[0] + 1});
   std::vector<int64_t> indices_shape({non_zero_length});
```

Comment:
Same question here about validating `indptr_shape` against `indptr_data` and `indices_data` against `indices_data...

Perhaps instead of calling the `Tensor` constructor directly, we need a Status-returning static factory function that validates input for consistency?

Reply:
That said, it seems we should then validate the indices _data_, which is much more expensive...

Reply:
You mean that it's better if we have Status-returning factory function for `Tensor` class and it validates indices data?

If so, it makes sense to me. SparseTensor should have Status-returning factory function, too.

Do you think we need to discuss on mailing-list before changing these APIs?

Reply:
> Do you think we need to discuss on mailing-list before changing these APIs?

I don't think so. We can keep the public (non-validating) constructors in addition to the (validating) factory functions. So there's no API breakage.

Reply:
I see.  I filed a new ticket https://issues.apache.org/jira/browse/ARROW-6508.